aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -0,0 +1,514 @@
+}
+
+SmallVector<char*, 4> TensorIterator::get_data_ptrs(ArrayRef<char*> base, IntList counter) const {
+  auto ptrs = SmallVector<char*, 4>(base);
+  for (int i = 0; i < ntensors(); i++) {
+    auto& stride = operands_[i].stride_;
+    for (int dim = 0; dim < ndim(); dim++) {
+      ptrs[i] += counter[dim] * stride[dim];
+    }
+  }
+  return ptrs;
+}
+
+SmallVector<char*, 4> TensorIterator::get_base_ptrs() const {
+  auto ptrs = SmallVector<char*, 4>();
+  for (int i = 0; i < ntensors(); i++) {
+    ptrs.push_back((char*)data_ptr(i));
+  }
+  return ptrs;
+}
+
+DimVector TensorIterator::make_counter(int64_t linear_offset) const {
+  auto counter = DimVector();
+  int64_t x = linear_offset;
+  for (auto size : shape_) {
+    counter.push_back(x % size);
+    x /= size;
+  }
+  AT_ASSERT(x == 0);
+  return counter;
+}
+
+void TensorIterator::increment_counter(DimVector& counter, int64_t n) const {
+  int64_t overflow = n;
+  for (int i = 0; i < ndim(); i++) {
+    auto size = shape_[i];
+    auto value = counter[i];
+    value += overflow;
+    overflow = value / size;
+    counter[i] = value % size;
+  }
+}
+
+void TensorIterator::for_each(loop_t loop) {
+  auto inner_strides = get_inner_strides();
+  auto base_ptrs = get_base_ptrs();
+
+  at::parallel_for(0, numel(), internal::GRAIN_SIZE, [&](int64_t begin, int64_t end) {
+    serial_for_each(loop, base_ptrs, inner_strides, begin, end - begin);
+  });
+}
+
+void TensorIterator::serial_for_each(loop_t loop, ArrayRef<char*> base_ptrs, IntList inner_strides, int64_t start, int64_t n) {
+  if (ndim() <= 1) {
+    auto ptrs = get_data_ptrs(base_ptrs, { start });
+    loop(ntensors(), ptrs.data(), inner_strides.data(), n);
+  } else {
+    auto counter = make_counter(start);
+    while (n > 0) {
+      auto ptrs = get_data_ptrs(base_ptrs, counter);
+      int64_t loop_size = std::min(n, shape_[0] - counter[0]);
+      loop(ntensors(), ptrs.data(), inner_strides.data(), loop_size);
+      n -= loop_size;
+      if (n == 0) break;
+      increment_counter(counter, loop_size);
+    }
+  }
+}
+
+bool TensorIterator::is_trivial_1d() const {
+  // TODO: check for casting once it's supported
+  return ndim() == 1;
+}
+
+bool TensorIterator::is_scalar(int arg) const {
+  const auto& stride = operands_[arg].stride_;
+  for (int i = 0; i < ndim(); i++) {
+    if (stride[i] != 0 && shape_[i] != 1) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool TensorIterator::is_cpu_scalar(int arg) const {
+  return is_scalar(arg) && operands_[arg].tensor_->type().backend() == at::kCPU;
+}
+
+void* TensorIterator::data_ptr(int arg) const {
+  return operands_[arg].data_;
+}
+
+void TensorIterator::remove_operand(int arg) {
+  operands_.erase(operands_.begin() + arg);
+}
+
+void TensorIterator::narrow(int dim, int64_t start, int64_t size) {
+  AT_ASSERT(dim < ndim() && size >= 1);
+  shape_[dim] = size;
+  for (auto& op : operands_) {
+    op.data_ = ((char*)op.data_) + op.stride_[dim] * start;
```

Comment:
do we check for empty tensors somewhere?  This math may or may not be correct for n-dimensional empty tensors.